File: sections/10_greedy_constructive/problems/b_bracket_completion/solution.cpp
#include <bits/stdc++.h>
using namespace std;

string complete(string s) {
    int n = (int)s.size();
    if (n % 2) return "IMPOSSIBLE";

    int need_open = n / 2;
    int need_close = n / 2;
    for (char ch : s) {
        if (ch == '(') --need_open;
        if (ch == ')') --need_close;
    }
    if (need_open < 0 || need_close < 0) return "IMPOSSIBLE";

    for (char &ch : s) {
        if (ch == '?') {
            if (need_open > 0) {
                ch = '(';
                --need_open;
            } else {
                ch = ')';
                --need_close;
            }
        }
    }

    int balance = 0;
    for (int i = 0; i < n; ++i) {
        balance += (s[i] == '(' ? 1 : -1);
        if (balance < 0) return "IMPOSSIBLE";
        if (i + 1 < n && balance == 0) return "IMPOSSIBLE";
    }
    if (balance != 0) return "IMPOSSIBLE";
    return s;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        string s;
        cin >> s;
        cout << complete(s) << '\n';
    }

    return 0;
}

File: sections/10_greedy_constructive/problems/b_bracket_completion/solution.py
"""Reference solution for B. Bracket Completion."""

import sys


def complete(s: str) -> str:
    n = len(s)
    if n % 2:
        return "IMPOSSIBLE"
    need_open = n // 2 - s.count("(")
    need_close = n // 2 - s.count(")")
    if need_open < 0 or need_close < 0:
        return "IMPOSSIBLE"

    chars = list(s)
    for i, ch in enumerate(chars):
        if ch == "?":
            if need_open > 0:
                chars[i] = "("
                need_open -= 1
            else:
                chars[i] = ")"
                need_close -= 1

    balance = 0
    for i, ch in enumerate(chars):
        balance += 1 if ch == "(" else -1
        if balance < 0:
            return "IMPOSSIBLE"
        if i + 1 < n and balance == 0:
            return "IMPOSSIBLE"
    if balance != 0:
        return "IMPOSSIBLE"
    return "".join(chars)


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    t = int(data[0])
    out: list[str] = []
    for i in range(1, t + 1):
        out.append(complete(data[i].decode()))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

